Throw an error when ambiguous sideset setting is requested.

roystgnr opened the pull request:
Refs #24597

## Reason
The error in #24597 ended up exhibiting as a nasty parallel-desynchronization issue that triggered stochastically.  #24619 removed the underlying error from the affected test, but this PR catches the desynchronization/stochastic stage of the bad behavior and replaces that with a deterministic error message.

## Design
Get all potential elements when looking for a sideset to flood, not just one.  Error if we find multiple candidate elements that would clearly lead to different sideset flood choices.

## Impact
This could cause an error to be thrown from existing input decks that are currently "mostly" working ... which is probably better than letting people continue to use such decks and having them quietly+intermittently give different or outright broken results.

roystgnr committed:
Detect ambiguity in SideSetsFromPointsGenerator

This appears to be the intermediate cause of the problem in #24597 - new
stitching behavior was leading to slits in the mesh, slits were leading
to ambiguous selection of elements/normals to flood from, and in
parallel if that ambiguity was resolved both ways, it was making libMesh
scream as soon as it noticed the inconsistent results.

roystgnr committed:
Add test for ambiguous-sideset error

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1574478/) on fd69577 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24624/site)

This comment will be updated on new commits.

lindsayad commented on test/tests/meshgenerators/sidesets_from_points_generator/tests:
```diff
@@ -9,5 +9,15 @@
+    design = 'meshgenerators/SideSetsFromPointsGenerator.md'
+    issues = '#24619'
+    recover = false
+  []
```
```suggestion
    mesh_mode = 'replicated'
  []
```
needed for stitched meshes

roystgnr replied:
Not since last year: https://github.com/libMesh/libmesh/pull/3262

roystgnr committed:
FillBetweenPointVectorsTools + DistributedMesh

I need to make another sweep through MOOSE, don't I?

As a rule of thumb:

"MeshSerializer(mesh)" says: "I can't write a distributed version of this
algorithm right now".  This is reasonable; sometimes that's crazy hard.

"ReplicatedMesh& mesh" says: "I won't even let you distribute the mesh
*later*".  This is unreasonable.  Learn to let go.

roystgnr committed:
FillBetweenSidesetsGenerator+DistributedMesh

I could have rewritten the unit test to not require this generator, but
better to just fix the problem.

roystgnr committed:
SideSetsFromPointsGenerator + DistributedMesh

Should I have called it "DistributableMesh", in hindsight?  I don't
think half our developers understand that the distribution isn't
*mandatory*, but keeping a mesh serialized while it's small and only
distributing it right before you extrude/refine/whatever to blow it up
is an excellent way to work.

lindsayad commented on framework/src/meshgenerators/SideSetsFromPointsGenerator.C:
```diff
@@ -72,28 +75,59 @@ SideSetsFromPointsGenerator::generate()
+                       normals[0]);
+
+          elem_to_flood = elem;
+          normal_to_flood = normals[0];
```
I guess even for curved meshes, it's probably a fine sanity check to just look at one of the normals

roystgnr replied:
The test vs `_variance` is looking at both normals; it's basically the same test `flood()` uses internally.  So if the test passes here then the flood should hit both; if it fails here then we have an actual ambiguity.

lindsayad commented on framework/src/utils/FillBetweenPointVectorsTools.C:
```diff
@@ -558,8 +563,12 @@ isBoundaryOpenSingleSegment(ReplicatedMesh & mesh,
 {
+  // This has no communication and expects elem_ptr to find any
+  // element, so it only works on serialized meshes
+  MeshSerializer serial(mesh);
```
is this a no-op on an already serialized mesh?

roystgnr replied:
Yup.

The one catch w.r.t. performance is that, if you're calling a bunch of functions like this in a row, you want the mesh to be already serialized, not rely on the serializer, because otherwise the serializer will happily serialize and then unserialize and serialize and unserialize as many times as you ask it to.

In the long run we might want some MOOSE interface for users to specify "don't bother distributing anything while we do all this stitching and such" vs "okay, we're about to extrude 100x more elements and then refine to get 8^n more, *definitely* make sure we're distributed first".  Maybe one MeshGenerator to turn off distribution and another to force it, or maybe an option in a base class parameter?

lindsayad commented on test/tests/meshgenerators/sidesets_from_points_generator/sidesets_ambiguity.i:
```diff
@@ -0,0 +1,350 @@
+[GlobalParams]
+  prevent_boundary_ids_overlap = false
```
I thought this had to be `true` to run into the issue?

lindsayad replied:
Otherwise how does the `preparedness` test work? Or did you modify the mesh generators? I guess I assumed this was a copy of that test

roystgnr replied:
Ah, shoot, since this is a copy of the preparedness test instead of a modification of it, you don't see the specific diffs highlighted, do you?  There's just one earlier but it's a big one:

```
      stitch_boundaries_pairs = '1   3;
                                 1   3;
                                 0   0' # 0 0 will leave a slit mesh
```

I'm not relying on auto-renumbering to tear the mesh here; I'm now deliberately leaving one of the input meshes unstitched.

lindsayad replied:
got ya 👍  I see that in there now

lindsayad commented on test/tests/meshgenerators/sidesets_from_points_generator/sidesets_ambiguity.i:
```diff
@@ -0,0 +1,350 @@
+  []
+[]
+
+[Variables]
```
I think everything from here on down can just be removed

roystgnr commented on framework/src/meshgenerators/SideSetsFromPointsGenerator.C:
```diff
@@ -72,28 +75,59 @@ SideSetsFromPointsGenerator::generate()
+
+          // If we *already* found a good but different side to paint
+          // our sideset with, we've got an ambiguity here.
+          if (elem_to_flood && (std::abs(1.0 - normal_to_flood * normals[0]) > _variance ||
```
Oh, sorry, I misunderstood.  Yeah, I'm being lazy here, but in my defense it's the same laziness `flood()` already uses; I'm not adding a new failure case.

Boy, now I feel more sympathy for all the people using `ReplicatedMesh` instead of `MeshSerializer`...

roystgnr committed:
Trim down new test to just the parts we need

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1577152/) on f0418cc wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ca2be3</th>
      <th colspan=4>#24624 f0418c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>85.41%</td>
      <td>+0.01%</td>
      <td>89.13%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88776</td>
      <td>88801</td>
      <td>+25</td>
      <td>41</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15172</td>
      <td>15164</td>
      <td>-8</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24624/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24624/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24624/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 89.13% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr committed:
Avoid unnecessary vector copies

Yeah, these are usually going to be small vectors, but let's set a good
example for future developers.

roystgnr committed:
Comment out overzealous warning

I normally don't like keeping commented out code in git, because that's
what the history is for, but "code that seems like it would be a good
idea plus an explanation of why it actually isn't" is an exception.

moosebuild commented:
Job [OpenMPI](https://civet.inl.gov/job/1580106/) on 26483b0 : invalidated by @lindsayad

lindsayad commented:
So was the fix commenting out the warning?

roystgnr commented:
Yup.  Looking through that code, there's basically no guarantee of one particular initial orientation.  Using a combination of generators that have simple element numberings and mesh settings that don't reorder ids must tend to prefer one orientation over another, but it's not something reliable and the automatic fix when needed isn't notable enough to bug the user with.
